eudemo/eudemo/ivc/panelling/_paneller.py

```diff
@@ -0,0 +1,252 @@
+        """Find z at the given normalised distance along the boundary."""
+        return self._z_spline(dist)
+
+    def x_tangent(self, dist: Union[float, np.ndarray]) -> np.ndarray:
+        """Find x at the tangent vector a given distance along the boundary."""
+        return self._x_tangent_spline(dist)
+
+    def z_tangent(self, dist: Union[float, np.ndarray]) -> np.ndarray:
+        """Find z at the tangent vector a given distance along the boundary."""
+        return self._z_tangent_spline(dist)
+
+
+def norm_lengths(points: np.ndarray) -> np.ndarray:
+    """
+    Calculate the cumulative normalized lengths between each 2D point.
+
+    Parameters
+    ----------
+    points
+        A numpy array of points, shape should be (2, N).
+    """
+    dists = np.diff(points, axis=1)
+    sq_dists = np.square(dists, out=dists)
+    summed_dists = np.sum(sq_dists, axis=0)
+    sqrt_dists = np.sqrt(summed_dists, out=summed_dists)
+    cumulative_sum = np.cumsum(summed_dists, out=sqrt_dists)
+    return np.hstack((0, cumulative_sum / cumulative_sum[-1]))
+
+
+def norm_tangents(points: np.ndarray) -> np.ndarray:
+    """
+    Calculate the normalised tangent vector at each of the given points.
+
+    Parameters
+    ----------
+    points
+        Array of coordinates. This must have shape (2, N), where N is
+        the number of points.
+
+    Returns
+    -------
+    tangents
+        The normalised vector of tangents.
+    """
+    grad = np.gradient(points, axis=1)
+    magnitudes = np.hypot(grad[0], grad[1])
+    return np.divide(grad, magnitudes, out=grad)
+
+
+def vector_intersect(p1, p2, p3, p4):
+    """
+    Find the point of intersection between two vectors defined by the given points.
+
+    Parameters
+    ----------
+    p1: np.ndarray(2)
+        The first point on the first vector
+    p2: np.ndarray(2)
+        The second point on the first vector
+    p3: np.ndarray(2)
+        The first point on the second vector
+    p4: np.ndarray(2)
+        The second point on the second vector
+
+    Returns
+    -------
+    p_inter: np.ndarray(2)
+        The point of the intersection between the two vectors
+    """
+    da = p2 - p1
+    db = p4 - p3
+
+    if np.isclose(np.cross(da, db), 0):  # vectors parallel
+        point = p2
+    else:
+        dp = p1 - p3
+        dap = normal_vector(da)
+        denom = np.dot(dap, db)
+        num = np.dot(dap, dp)
+        point = num / denom.astype(float) * db + p3
+    return point
+
+
+def normal_vector(side_vectors):
+    """
+    Anti-clockwise
+
+    Parameters
+    ----------
+    side_vectors: np.array(N, 2)
+        The side vectors of a polygon
+
+    Returns
+    -------
+    a: np.array(2, N)
+        The array of 2-D normal vectors of each side of a polygon
+    """
+    a = -np.array([-side_vectors[1], side_vectors[0]]) / np.sqrt(
+        side_vectors[0] ** 2 + side_vectors[1] ** 2
+    )
+    a[np.isnan(a)] = 0
```

Comment:
Have you seen any divide by NaN warnings come up here? It seems possible at any rate

Reply:
Do you mean divide by 0 warnings from calculating `a`? I haven't come across any in my testing. I'll admit this is ripped straight from BLUEPRINT, so I'm fuzzy on the justification for this line

Reply:
I did, sure, if you havent seen any that fine

Reply:
This is another one that was already in the `geometry` module https://github.com/Fusion-Power-Plant-Framework/bluemira/pull/2060/files#diff-fddce791a384b13641d79f8580424a09ee09d2f3def690f575c9ba95d67fb44eL101